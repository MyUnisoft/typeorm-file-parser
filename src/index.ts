// Import Node.js Dependencies
import { createReadStream } from "fs";

// Import Third-party Dependencies
import split2 from "split2";

// Import Internal Dependencies
import { parseDecorator, TypeORMDecoratorBase, TypeORMDecoratorExtended } from "./decorator/parser";

// CONSTANTS
const kArrobaseChar = "@";
const kSlashChar = "/";

export async function* lazyFetchFileDecorator(fileLocation: string) {
  const rStream = createReadStream(fileLocation).pipe(split2());
  let pastConcat: string | null = null;

  for await (const line of rStream) {
    const trimedLine = line.trim() as string;
    if (trimedLine.includes("export interface")) {
      break;
    }

    const isCommentLine = trimedLine.startsWith(kSlashChar);
    if (isCommentLine) {
      continue;
    }

    const fieldLineResult = /^([a-zA-Z0-9_]+)\s*:\s*([a-zA-Z]+)/g.exec(trimedLine);
    if (fieldLineResult !== null) {
      const [, name, value] = fieldLineResult;
      yield [name, value];

      continue;
    }

    const isEndLine = trimedLine.charAt(trimedLine.length - 1) === ")";
    const isStartLine = trimedLine.startsWith(kArrobaseChar);

    if (isEndLine) {
      const finalPayload = isStartLine ? trimedLine : pastConcat + trimedLine;
      yield parseDecorator(finalPayload);

      pastConcat = null;
      continue;
    }

    pastConcat = (pastConcat || "") + trimedLine;
  }
}

type DecoratorExWithoutName = Omit<TypeORMDecoratorExtended, "name"> | string;

function removeDecoratorName(decorator: TypeORMDecoratorBase): DecoratorExWithoutName {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { name, ...otherProperties } = decorator;
  if (Object.keys(otherProperties).length === 0) {
    return name;
  }

  return otherProperties;
}

export interface TypeORMProperty {
  type: string;
  decorators: Record<string, TypeORMDecoratorBase>;
}

export interface ParsedTypeORMResult {
  unique?: DecoratorExWithoutName;
  properties: Record<string, TypeORMProperty>;
}

export async function readFile(fileLocation: string): Promise<ParsedTypeORMResult> {
  let memoryColumns: TypeORMDecoratorBase[] = [];

  const result: ParsedTypeORMResult = {
    properties: {}
  };
  const asyncIter = await lazyFetchFileDecorator(fileLocation);
  await asyncIter.next();

  for await (const decorator of asyncIter) {
    if (Array.isArray(decorator)) {
      const decorators = {};
      for (const decorator of memoryColumns) {
        decorators[decorator.name] = decorator;
      }

      const [propertyName, propertyValue] = decorator;
      result.properties[propertyName] = {
        type: propertyValue,
        decorators
      };

      memoryColumns = [];
    }
    else if (decorator.name === "Unique") {
      result.unique = removeDecoratorName(decorator);
    }
    else {
      memoryColumns.push(decorator);
    }
  }

  return result;
}
