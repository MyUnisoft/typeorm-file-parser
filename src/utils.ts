export function charSet(...plages: (string | [number, number])[]): Set<string> {
  const ret = new Set<string>();

  for (const plage of plages) {
    if (typeof plage === "string") {
      ret.add(plage);
      continue;
    }

    for (let tid = plage[0]; tid <= plage[1]; tid++) {
      ret.add(String.fromCharCode(tid));
    }
  }

  return ret;
}

export const END_OF_SEQUENCE = Symbol("EOS");
export type SEQUENCE_OR_TOKEN<T> = typeof END_OF_SEQUENCE | T;

export function getNextItem<T>(iterator: IterableIterator<T>): SEQUENCE_OR_TOKEN<T> {
  const item = iterator.next();

  return item.done ? END_OF_SEQUENCE : item.value;
}
