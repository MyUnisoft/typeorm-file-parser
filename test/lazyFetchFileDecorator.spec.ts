// Import Node.js Dependencies
import path from "path";

// Import Internal Dependencies
import { lazyFetchFileDecorator } from "../src/index";

// CONSTANTS
const kFixturesPath = path.join(__dirname, "fixtures");

describe("lazyFetchFileDecorator", () => {
  it("should fetch all properties and decorators from GiArticle", async() => {
    const result: any[] = [];
    const asyncIter = await lazyFetchFileDecorator(path.join(kFixturesPath, "GiArticle.ts"));

    for await (const decorator of asyncIter) {
      result.push(decorator as any);
    }

    expect(result).toMatchSnapshot("GiArticleFixture");
  });
});

