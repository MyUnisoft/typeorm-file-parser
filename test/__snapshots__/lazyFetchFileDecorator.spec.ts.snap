// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lazyFetchFileDecorator should fetch all properties and decorators from GiArticle: GiArticleFixture 1`] = `
Array [
  Object {
    "name": "Entity",
  },
  Object {
    "columns": Array [
      "reference",
    ],
    "constraintName": "GiArticle_reference",
    "name": "Unique",
  },
  Object {
    "name": "PrimaryGeneratedColumn",
  },
  Array [
    "id",
    "number",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "length": 50,
    },
    "type": "varchar",
  },
  Array [
    "name",
    "string",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "length": 6,
    },
    "type": "varchar",
  },
  Array [
    "reference",
    "string",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "nullable": true,
    },
    "type": "text",
  },
  Array [
    "note",
    "string",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "enum": "giTypeArticle",
    },
    "type": "enum",
  },
  Array [
    "type",
    "giTypeArticle",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "enum": "periodicity",
    },
    "type": "enum",
  },
  Array [
    "periodicity",
    "periodicity",
  ],
  Object {
    "name": "Column",
    "properties": Object {},
    "type": "boolean",
  },
  Array [
    "mandatoryArticle",
    "boolean",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "enum": "giExercice",
    },
    "type": "enum",
  },
  Array [
    "exercice",
    "string",
  ],
  Object {
    "name": "Column",
    "properties": Object {},
    "type": "smallint",
  },
  Array [
    "firstDeadline",
    "number",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "length": 6,
    },
    "type": "varchar",
  },
  Array [
    "firmAccount",
    "number",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "length": 6,
    },
    "type": "varchar",
  },
  Array [
    "firmCollectedVat",
    "string",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "length": 6,
    },
    "type": "varchar",
  },
  Array [
    "clientAccount",
    "string",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "length": 6,
    },
    "type": "varchar",
  },
  Array [
    "clientDeductibleVat",
    "string",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "default": "false",
    },
    "type": "boolean",
  },
  Array [
    "billed",
    "boolean",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "default": "false",
    },
    "type": "boolean",
  },
  Array [
    "archived",
    "boolean",
  ],
  Object {
    "name": "Column",
    "properties": Object {
      "enum": "giAmount",
    },
    "type": "enum",
  },
  Array [
    "amount",
    "giAmount",
  ],
  Object {
    "name": "ManyToMany",
    "properties": Object {},
    "table": "GiMission",
    "tableColumn": "articles",
  },
  Object {
    "name": "JoinTable",
    "properties": Object {
      "name": "gi_articles_missions",
    },
  },
  Array [
    "missions",
    "GiMission",
  ],
  Object {
    "name": "OneToOne",
    "properties": Object {},
    "table": "GiArticle",
    "tableColumn": "id",
  },
  Object {
    "name": "JoinColumn",
    "properties": Object {
      "name": "account_article_id",
    },
  },
  Array [
    "accountArticle",
    "GiArticle",
  ],
  Object {
    "name": "ManyToOne",
    "properties": Object {
      "cascade": true,
      "nullable": false,
    },
    "table": "WorkUnit",
    "tableColumn": "articles",
  },
  Array [
    "workUnit",
    "WorkUnit",
  ],
  Object {
    "name": "ManyToOne",
    "properties": Object {
      "nullable": false,
    },
    "table": "Vat",
    "tableColumn": "giArticlesFirm",
  },
  Object {
    "name": "JoinColumn",
    "properties": Object {
      "name": "firm_vat_id",
    },
  },
  Array [
    "firmVat",
    "Vat",
  ],
  Object {
    "name": "ManyToOne",
    "properties": Object {
      "nullable": false,
    },
    "table": "Vat",
    "tableColumn": "giArticlesClient",
  },
  Object {
    "name": "JoinColumn",
    "properties": Object {
      "name": "client_vat_id",
    },
  },
  Array [
    "clientVat",
    "Vat",
  ],
  Object {
    "name": "OneToMany",
    "properties": Object {
      "cascade": true,
    },
    "table": "GiVintage",
    "tableColumn": "article",
  },
  Array [
    "vintages",
    "GiVintage",
  ],
]
`;
